### ISSUE TITLE
Modified delayed_malloc to work on arrays, doesn't work with current vedaMemFree/Copy & context sync order.

### ISSUE THREAD

**Issue by ftpronk** [SX-Aurora/veda#6]
I have been using the delayed_malloc as an inspiration to try out my own simple test.  Instead of copying a char array, I've been copying two vectors/arrays that are added in a kernel into a third array. 

When delay-allocating this third array on the VE, and keeping the memCopy, memFree & ctx synchronization in the same order as in the example case, everything seems to be working fine. So the order is:
```
VEDA(vedaMemcpyDtoHAsync(cpu_return, kern_sum, N*sizeof(int), stream));

VEDA(vedaMemFreeAsync(x_device, stream));
VEDA(vedaMemFreeAsync(y_device, stream));
VEDA(vedaMemFreeAsync(kern_sum,  stream));

vedaCtxSynchronize();
```

Now. If I replace the device allocation by a pre-allocation, then the array returned has some garbage in the first few elements.  The only way to get a proper data return is to put the `vedaCtxSynchronize()` call before the `vedaMemFreeAsync()` calls.

Did I do something wrong, or is this a bug?  Intuitively I would have put a sync() before a free() to start with, but perhaps that's then missing the point of the `vedaMemFreeAsync()` functionality?

See attachment for a minimal test case.

[simple_add_kernel.tar.gz](https://github.com/SX-Aurora/veda/files/5533244/simple_add_kernel.tar.gz)

**Comment by mergian** [comment]
I can't reproduce your problem :( Have you tried the newest version v0.9.5? There was a problem in the previous AVEO that could cause wrong arguments being copied to the VE.

I get following output:
```
device sum:
5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5

host after copy:
5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5

The sum returned by the kernel is correct
```

**Comment by ftpronk** [comment]
I forgot to add an output of my own run, sorry! This is what I get:
```
device sum:
5 5 5 5 5 5 5 5 5 5 

host after copy:
16021648 24592 5 5 5 5 5 5 5 5 
```
As the device sum printed is correct, I don't think it's a copying problem to the VE. Those values come from the array that is being copied back, so it would rather be a copying problem to the VH.

As it stands we are using veda version `0.9.2` if I believe the folder name in `/usr/local/ve`.  I'll ask Erich if he can upgrade the aurora machine.

**Comment by ftpronk** [comment]
Using `vedaMemcpyDtoH()` instead of `vedaMemcpyDtoHAsync()` works correctly.

**Comment by mergian** [comment]
That is expected as the non async method explicitly call sync on the context after issuing the copy command. I still think that the problem is solved in the v0.9.5. Please let us know when you had been able to test it with the new release.

**Comment by ftpronk** [comment]
The 0.9.5. release seems to have solved this issue, so that is good news. I'll close this issue.
